File: js/test/stack.js
var test = require('tape');
/*
 * STACKS ARE LAST IN FIRST OUT.
 */


var Item = function(value, next) {
  this.value = value;
  this.next  = next;
}

var Stack = function() {
  this.top = undefined;
}

Stack.prototype.push = function(value) {
  if (this.top === undefined) {
    this.top = new Item(value);
  } else {
    this.top = new Item(value, this.top);
  }
}

Stack.prototype.peek = function() {
  return (this.top !== undefined) ? this.top.value : undefined;
}

Stack.prototype.pop = function() {
  if (this.top === undefined) {
    return undefined;
  } else {
    var value = this.top.value;
    this.top  = this.top.next;
    return value;
  }
}


test('testStack', function (t) {
  t.plan(10);

  var stack = new Stack();
  t.equal(stack.peek(), undefined);
  t.equal(stack.pop(),  undefined);

  stack.push(1);
  t.equal(stack.peek(), 1);
  t.equal(stack.pop(),  1);
  t.equal(stack.pop(),  undefined);

  stack.push(1);
  stack.push(2);
  stack.push(3);
  t.equal(stack.pop(),  3);
  t.equal(stack.peek(), 2);
  t.equal(stack.pop(),  2);
  t.equal(stack.pop(),  1);
  t.equal(stack.pop(),  undefined);
});

File: java/src/main/java/org/anhonesteffort/algosndatas/Stack.java
/*
 * Copyright (C) 2016 An Honest Effort LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.anhonesteffort.algosndatas;

/*
  STACKS ARE LAST IN FIRST OUT.
 */
public class Stack<T> {

  private Item<T> top;

  public void push(T value) {
    if (top == null) {
      top = new Item<>(value, null);
    } else {
      top = new Item<>(value, top);
    }
  }

  public T peek() {
    return (top != null) ? top.value : null;
  }

  public T pop() {
    if (top != null) {
      T value = top.value;
      top = top.next;
      return value;
    } else {
      return null;
    }
  }

  private class Item<V extends T> {
    private final V value;
    private final Item<V> next;

    public Item(V value, Item<V> next) {
      this.value = value;
      this.next  = next;
    }
  }

}
